engine/src/main/java/io/camunda/zeebe/engine/processing/streamprocessor/ProcessingStateMachine.java

```diff
@@ -167,45 +166,50 @@ public ProcessingStateMachine(
     this.shouldProcessNext = shouldProcessNext;
 
     final int partitionId = logStream.getPartitionId();
-    typedEvent = new TypedEventImpl(partitionId);
+    typedCommand = new TypedEventImpl(partitionId);
     responseWriter = context.getWriters().response();
 
     metrics = new StreamProcessorMetrics(partitionId);
     streamProcessorListener = context.getStreamProcessorListener();
   }
 
   private void skipRecord() {
-    notifySkippedListener(currentEvent);
-    actor.submit(this::readNextEvent);
+    notifySkippedListener(currentRecord);
+    actor.submit(this::readNextRecord);
     metrics.eventSkipped();
   }
 
-  void readNextEvent() {
+  void readNextRecord() {
     if (onErrorRetries > 0) {
       onErrorHandlingLoop = false;
       onErrorRetries = 0;
     }
 
-    tryToReadNextEvent();
+    tryToReadNextRecord();
   }
 
-  private void tryToReadNextEvent() {
+  private void tryToReadNextRecord() {
     final var hasNext = logStreamReader.hasNext();
 
-    if (currentEvent != null) {
+    if (currentRecord != null) {
+      final var previousRecord = currentRecord;
       // All commands cause a follow-up event or rejection, which means the processor
       // reached the end of the log if:
       //  * the last record was an event or rejection
       //  * and there is no next record on the log
-      final var previousEvent = currentEvent;
-      reachedEnd = commandFilter.applies(previousEvent) && !hasNext;
+      //  * and this was the last record written (records that have been written to the dispatcher
+      //    might not be written to the log yet, which means they will appear shortly after this)
+      reachedEnd =
+          commandFilter.applies(previousRecord)
+              && !hasNext
+              && lastWrittenPosition == previousRecord.getPosition();
```

Comment:
:wrench: This check should work in most cases. However, we could make the condition a bit more flexible by supporting the case that the last record was not written by this stream processor. 

```suggestion
              && lastWrittenPosition <= previousRecord.getPosition();
```

Reply:
👍  Good point